tempesta_fw/http_parser.c

```diff
@@ -1764,19 +1765,39 @@ __parse_transfer_encoding(TfwHttpMsg *hm, unsigned char *data, size_t len)
 		__FSM_I_JMP(I_TransEncodOther);
 	}
 
+
+	/*
+	 * RFC 7230 3.3.1:
+	 * If any transfer coding
+	 * other than chunked is applied to a REQUEST payload body, the sender
+	 * MUST apply chunked as the final transfer coding to ensure that the
+	 * message is properly framed. If any transfer coding other than
+	 * chunked is applied to a RESPONSE payload body, the sender MUST either
+	 * apply chunked as the final transfer coding or terminate the message
+	 * by closing the connection.
+	 *
+	 * TODO: process transfer encodings: gzip, deflate, identity,
+	 * compress;
+	 */
 	__FSM_STATE(I_TransEncodOther) {
-		/*
-		 * TODO: process transfer encodings: gzip, deflate, identity,
-		 * compress;
-		 */
 		__FSM_I_MATCH_MOVE(token, I_TransEncodOther);
 		c = *(p + __fsm_sz);
 		if (IS_WS(c) || c == ',')
 			__FSM_I_MOVE_n(I_EoT, __fsm_sz + 1);
 		if (IS_CRLF(c)) {
-			/* "chunked" must be the last coding. */
 			if (unlikely(test_bit(TFW_HTTP_B_CHUNKED, msg->flags)))
-				return CSTR_NEQ;
+			{
+				if (client)
+					return CSTR_NEQ;
+				/*
+				 * Prepare to receive 'unlimited' body, close
+				 * client connection after response is forwarded
+				 * to preserve correct message framing on client
+				 * side.
+				 */
+				__clear_bit(TFW_HTTP_B_CHUNKED, msg->flags);
+				set_bit(TFW_HTTP_B_CONN_CLOSE, msg->pair->flags);
```

Comment:
We get to this line only for responses, and only where `Transfer-Encoding` had a "chunked" somewhere in the field value. So, as far as I understand, if backend sends us `Transfer-Encoding: gzip`, we won't be ready for the "unlimited body mode", since there were no "chunked", and therefore TFW_HTTP_B_CHUNKED bit wasn't set. Right?

Reply:
The "unlimited body mode" is turned on in two cases: https://github.com/tempesta-tech/tempesta/blob/1d9a5758b966ed93a78ae4264a592f67023e68a2/tempesta_fw/http_parser.c#L1022

1. Transfer-Encoding header is not empty and the `CHUNKED` flag is missing
2. No Transfer-Encoding and no Content-Length headers are provided. 

So if backend sends us Transfer-Encoding: gzip, we will be ready for the "unlimited body mode".

But there is an issue in the PR: `TFW_HTTP_B_UNLIMITED` flag is not set, so the response is actually blocked and dropped after receive. I will need to move `TFW_HTTP_B_UNLIMITED` flag to the `Resp_BodyUnlimStart` parser state.

Reply:
Thanks to pointing to that. I have reworked the PR, the comment below provides the explanations about flags and their usage.